sound/soc/sof/intel/hda-dai.c

```diff
@@ -56,13 +63,22 @@ hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai
 	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(cpu_dai->component);
 	struct snd_sof_widget *swidget = w->dobj.private;
-	struct snd_sof_dai *sdai = swidget->private;
+	const struct hda_dai_widget_dma_ops *ops;
+	struct snd_sof_dai *sdai;
+
+	if (!swidget) {
+		ops = hda_select_dai_widget_ops(sdev, NULL);
+		if (!ops || !ops->get_hext_stream || !ops->setup_hext_stream)
```

Comment:
why is setup_hext_stream mandatory now?

Reply:
I don't think we can have different set of mandatory ops for different instances of the ops. setup() is optional for this exact reason.

Reply:
This is exactly what I was saying for the Abstraction PR. This two op is mandatory for normal mode and for DSPless.
In fact, the hext_stream ops are mandatory for the normal mode.
I can drop the mandatory check, but it is mandatory.

And the fact is that different modes (normal chain and DSPless) very well needs different sets of mandatory callbacks.

Reply:
@ujfalusi different modes needing different ops doesnt make the ops mandatory. Its a bit like saying the IPC ops we have for IPC3/IPC4 will need to have different set of mandatory ops right? But if you think setup is mandatory for all cases, lets make it mandatory